### ISSUE TITLE
different ranks are trying to submit different tensors

### ISSUE THREAD

**Issue by XuanBaby** [horovod/horovod#1361]
**Environment:**
1. Framework: (TensorFlow, Keras, PyTorch, MXNet)：Tensorflow
2. Framework version:1.13
3. Horovod version:horovod-0.16.2-py3.6
4. MPI version:4.0.0
5. CUDA version:AMD
6. NCCL version:
7. Python version:3.6.8
8. OS and version:centos 7.4
9. GCC version:7.2.1

**Checklist:**
1. Did you search issues to find if somebody asked this question before?
2. If your question is about hang, did you read [this doc](https://github.com/horovod/horovod/blob/master/docs/running.rst)?
3. If your question is about docker, did you read [this doc](https://github.com/horovod/horovod/blob/master/docs/docker.rst)?
4. Did you check if you question is answered in the [troubleshooting guide](https://github.com/horovod/horovod/blob/master/docs/troubleshooting.rst)?

**Your question:**
[2019-08-28 16:25:17.568806: W horovod/common/operations.cc:764] One or more tensors were submitted to be reduced, gathered or broadcasted by subset of ranks and are waiting for remainder of ranks for more than 60 seconds. This may indicate that different ranks are trying to submit different tensors or that only subset of ranks is submitting tensors, which will cause deadlock.
what should I do  for this reminder?

**Comment by XuanBaby** [comment]
Hi, @tgaddair thank you for your answer,as your advice, I test the /horovod/example/tensorflow_synthetic_benchmark.py :
mpirun -np 2 --allow-run-as-root python3 tensorflow_synthetic_benchmark.py
the output:
2019-08-30 15:00:23.412460: I tensorflow/core/kernels/conv_grad_filter_ops.cc:979] running auto-tune for Backward-Filter
2019-08-30 15:00:23.423146: I tensorflow/core/kernels/conv_grad_filter_ops.cc:979] running auto-tune for Backward-Filter
Running benchmark...
Iter #0: 1.8 img/sec per GPU
Iter #1: 1.8 img/sec per GPU
Iter #2: 1.8 img/sec per GPU

**Comment by XuanBaby** [comment]
@tgaddair I could not solve this question,can you help me? or give me some ideal to find the reason? Thank you very much!

**Comment by abditag2** [comment]
It seems that you can run the benchmarks which indicate that the problem is most likely in your training script. How are you dividing the data between your workers? You need to make sure that each worker processes "exactly" the same number of samples of data.

**Comment by XuanBaby** [comment]
@tgaddair I just run "mpirun -np 2 --allow-run-as-root python3 tensorflow_synthetic_benchmark.py" without other args.I use strace -p PID,get messages:
futex(0x6c0dad8, FUTEX_WAIT_BITSET_PRIVATE|FUTEX_CLOCK_REALTIME, 0, NULL, 0xffffffff) = 0
write(1, "    80     1     7.7   9.314 0.0"..., 39) = 39
futex(0x6d1fab4, FUTEX_WAKE_OP_PRIVATE, 1, 1, 0x6d1fab0, FUTEX_OP_SET<<28|0<<12|FUTEX_OP_CMP_GT<<24|0x1) = 1
futex(0x6d1f9b4, FUTEX_WAKE_OP_PRIVATE, 1, 1, 0x6d1f9b0, FUTEX_OP_SET<<28|0<<12|FUTEX_OP_CMP_GT<<24|0x1) = 1
futex(0x6c0dad8, FUTEX_WAIT_BITSET_PRIVATE|FUTEX_CLOCK_REALTIME, 0, NULL, 0xffffffff) = 0
futex(0x6d1fab4, FUTEX_WAKE_OP_PRIVATE, 1, 1, 0x6d1fab0, FUTEX_OP_SET<<28|0<<12|FUTEX_OP_CMP_GT<<24|0x1) = 1
futex(0x6d1f9b4, FUTEX_WAKE_OP_PRIVATE, 1, 1, 0x6d1f9b0, FUTEX_OP_SET<<28|0<<12|FUTEX_OP_CMP_GT<<24|0x1) = 1
futex(0x6c0dad8, FUTEX_WAIT_BITSET_PRIVATE|FUTEX_CLOCK_REALTIME, 0, NULL, 0xffffffff) = 0
futex(0x6d1fab4, FUTEX_WAKE_OP_PRIVATE, 1, 1, 0x6d1fab0, FUTEX_OP_SET<<28|0<<12|FUTEX_OP_CMP_GT<<24|0x1) = 1
I consider the reason why program hang is it wait to communication between process ,But I'm not sure what it's waiting for,How can I do for this situation?

**Comment by Himscipy** [comment]
Hi,
I am facing the same issue, I have a data-pipeline similar to one of many examples of Tensorflow.
as following
`ef build_input_pipeline(x_train, x_test, y_train, y_test,
                         batch_size, valid_size,SEED=None):
    """Build an Iterator switching between train and heldout data."""

    x_train = x_train.astype("float32")
    x_test = x_test.astype("float32")

    x_train /= 255
    x_test /= 255

    y_train = y_train.flatten()
    y_test = y_test.flatten()

    if FLAGS.subtract_pixel_mean:
        x_train_mean = np.mean(x_train, axis=0)
        x_train -= x_train_mean
        x_test -= x_train_mean

    #print("x_train shape:" + str(x_train.shape))
    #print(str(x_train.shape[0]) + " train samples")
    #print(str(x_test.shape[0]) + " test samples")

    # Build an iterator over training batches.
    training_dataset = tf.data.Dataset.from_tensor_slices(
        (x_train, np.int32(y_train)))
    training_batches = training_dataset.shuffle(
        50000, reshuffle_each_iteration=True, seed=SEED() ).repeat().batch(batch_size)
    training_iterator = tf.data.make_one_shot_iterator(training_batches).string_handle()

    # Build a iterator over the heldout set with batch_size=heldout_size,
    # i.e., return the entire heldout set as a constant.
    heldout_dataset = tf.data.Dataset.from_tensor_slices(
        (x_test, np.int32(y_test)))
    heldout_batches = heldout_dataset.repeat().batch(valid_size)
    heldout_iterator = tf.data.make_one_shot_iterator(heldout_batches).string_handle()

    # Combine these into a feedable iterator that can switch between training
    # and validation inputs.
    handle = tf.placeholder(tf.string, shape=[])
    feedable_iterator = tf.data.Iterator.from_string_handle(
        handle, training_batches.output_types, training_batches.output_shapes)
    images, labels = feedable_iterator.get_next()
    print ('Done with datapipeline')
    return images, labels, handle, training_iterator, heldout_iterator`

I then use this in the main function as given below, since each hvd.init is called this pipeline routine will be executed by every rank therefore I have used the seed argument so that each rank can have different handle generated.@abditag2 do you think this is a right way of feeding data using horovod.

`
def main(argv):
    
    hvd.init()
    
    config = create_config_proto()
    Seed_set = Setup_Seed()

    print("Total Number of workers",hvd.size())
    if FLAGS.fake_data:
        (x_train, y_train), (x_test, y_test) = build_fake_data()
    else:
        # TODO Load MNIST data
        # TODO Load MNIST-80 
        # Load CIFAR-10 data
        (x_train, y_train), (x_test, y_test) = load_data()

        # Uncomment if using this code locally.
        # (x_train, y_train), (x_test, y_test) = tf.keras.datasets.cifar10.load_data()

    Time_data_set_start = time.time()
    (images, labels, handle,training_iterator,
    heldout_iterator) = build_input_pipeline(x_train, x_test, y_train, y_test,
                                                FLAGS.batch_size, 500, SEED=Seed_set)
      ;
      :
      with tf.train.MonitoredTrainingSession(checkpoint_dir=checkpoint_dir,hooks=hooks,config=config) as sess:
        
        # Uncomment to see model Parameters

        # if hvd.rank() == 0:
        #     print(model.summary())

        train_handle, heldout_handle  = sess._coordinated_creator.tf_sess.run([training_iterator,heldout_iterator])

        # Run the training loop
        # Time start
        start_Train_time = time.time()
        Time_to_train = 0

        while not sess.should_stop():

            step = sess._coordinated_creator.tf_sess.run(global_step)

            start_batch_time = time.time()
            
            _ = sess.run([train_op,train_accuracy_update_op,update_step_op],feed_dict={handle: train_handle})

`

**Comment by iou2much** [comment]
same issue here.. does anyone manage it solve this? seems really tricky one
